File: leetcode/most-stones-removed-with-same-row-or-column/most-stones-removed-with-same-row-or-column.go
package main

import (
	"fmt"
)

// source: https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/

func removeStones(stones [][]int) int {
	n := len(stones)
	parent := make([]int, n)
	xs, ys := make(map[int][]int), make(map[int][]int)
	var find func(int) int
	var union func(int, int)

	union = func(x, y int) {
		parent[find(x)] = find(y)
	}
	find = func(x int) int {
		if parent[x] != x {
			parent[x] = find(parent[x])
		}
		return parent[x]
	}

	for i, s := range stones {
		parent[i] = i
		x, y := s[0], s[1]
		if _, ok := xs[x]; !ok {
			xs[x] = make([]int, 0)
		}
		if _, ok := ys[y]; !ok {
			ys[y] = make([]int, 0)
		}
		xs[x] = append(xs[x], i)
		ys[y] = append(ys[y], i)
	}

	for _, ps := range xs {
		s := ps[0]
		for i := 1; i < len(ps); i++ {
			union(s, ps[i])
		}
	}
	for _, ps := range ys {
		s := ps[0]
		for i := 1; i < len(ps); i++ {
			union(s, ps[i])
		}
	}

	res := make(map[int]bool, 0)
	for i := 0; i < n; i++ {
		res[find(i)] = true
	}

	return n - len(res)
}

func main() {
	testCases := []struct {
		stones [][]int
		want   int
	}{
		{
			stones: [][]int{{1, 2}, {1, 3}, {3, 3}, {3, 1}, {2, 1}, {1, 0}},
			want:   5,
		},
		{
			stones: [][]int{{0, 0}, {0, 1}, {1, 0}, {1, 2}, {2, 1}, {2, 2}},
			want:   5,
		},
		{
			stones: [][]int{{0, 0}, {0, 2}, {1, 1}, {2, 0}, {2, 2}},
			want:   3,
		},
		{
			stones: [][]int{{0, 0}},
			want:   0,
		},
	}

	successes := 0
	for _, tc := range testCases {
		x := removeStones(tc.stones)
		status := "ERROR"
		if fmt.Sprint(x) == fmt.Sprint(tc.want) {
			status = "OK"
			successes++
		}
		fmt.Println(status, "	Expected: ", tc.want, " Actual: ", x)
	}
	if l := len(testCases); successes == len(testCases) {
		fmt.Printf("===\nSUCCESS: %d of %d tests ended successfully\n", successes, l)
	} else {
		fmt.Printf("===\nFAIL: only %d of %d tests ended successfully\n", successes, l)
	}

}

File: leetcode/most-stones-removed-with-same-row-or-column/most-stones-removed-with-same-row-or-column.py
# source: https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/
from typing import List


class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        n = len(stones)
        ds = dict()
        rows, cols = dict(), dict()

        def union(x, y):
            ds[find(x)] = find(y)

        def find(x):
            if ds[x] != x:
                ds[x] = find(ds[x])
            return ds[x]

        for i, s in enumerate(stones):
            ds[(s[0], s[1])] = (s[0], s[1])

            if s[0] not in rows:
                rows[s[0]] = list()
            rows[s[0]].append(i)

            if s[1] not in cols:
                cols[s[1]] = list()
            cols[s[1]].append(i)

        for k, v in rows.items():
            for c in v:
                union((stones[c][0], stones[c][1]), (stones[v[0]][0], stones[v[0]][1]))

        for k, v in cols.items():
            for c in v:
                union((stones[c][0], stones[c][1]), (stones[v[0]][0], stones[v[0]][1]))

        groups = set()

        for i in stones:
            if find((i[0], i[1])) not in groups:
                groups.add(find((i[0], i[1])))

        return n - len(groups)


sol = Solution()
sol.removeStones(stones=[[0, 1], [1, 2], [1, 3], [3, 3], [2, 3], [0, 2]])
sol.removeStones(stones=[[0, 0], [0, 1], [1, 0], [1, 2], [2, 1], [2, 2]])
sol.removeStones(stones=[[0, 0], [0, 2], [1, 1], [2, 0], [2, 2]])
